sklearn/neighbors/loop.py

```diff
@@ -0,0 +1,447 @@
+    def _prob_local_outlier_factors(prob_set_distances_X,
+                                    prob_set_distances_ev_X):
+        """
+        Calculates the probabilistic local outlier factor of each sample. As
+        this method is used in the calculation of the local probability of
+        each sample, this method is kept private.
+
+        Parameters
+        ----------
+        prob_set_distances_X : array-like, shape (n_samples,)
+            The probabilistic distances of X.
+
+        prob_set_distances_ev_X : array-like, shape (n_samples,)
+            The expected value of the probabilistic distances of X.
+
+        Returns
+        -------
+        prob_local_outlier_factors : array, shape (n_samples,)
+            The probabilistic local outlier factor of each sample.
+        """
+        prob_local_outlier_factors \
+            = (prob_set_distances_X / prob_set_distances_ev_X) - 1.
+
+        return prob_local_outlier_factors
+
+    @staticmethod
+    def _prob_local_outlier_factors_ev(prob_local_outlier_factors_X):
+        """
+        Calculates the expected value of the probabilistic local outlier
+        factor. As this method is used in the calculation of the local
+        probability of each sample, this method is kept private.
+
+        Parameters
+        ----------
+        prob_local_outlier_factors_X : array-like, shape (n_samples,)
+            The probabilistic local outlier factor of each sample.
+
+        Returns
+        -------
+        prob_local_outlier_factors_ev : array, shape (n_samples,)
+            The expected value of the probabilistic local outlier factor of X
+            as an array of repeated values.
+        """
+        prob_local_outlier_factors_ev = np.array([np.sum(
+            np.power(prob_local_outlier_factors_X[
+                         ~np.isinf(prob_local_outlier_factors_X)],
+                     2) /
+            prob_local_outlier_factors_X.shape[0])] *
+                                         prob_local_outlier_factors_X.shape[
+                                             0])
+
+        return prob_local_outlier_factors_ev
+
+    def _norm_prob_local_outlier_factors(self,
+                                         prob_local_outlier_factors_ev_X):
+        """
+        Calculates the normalized probabilistic local outlier factor. As this
+        method is used in the calculation of the local probability of each
+        sample, this method is kept private.
+
+        Parameters
+        ----------
+        prob_local_outlier_factors_ev_X : array-like, shape (n_samples,)
+            The expected value of the probabilistic local outlier factor of X
+            as an array of repeated values.
+
+        Returns
+        -------
+        norm_prob_local_outlier_factors : array, shape (n_samples,)
+            The normalized probabilistic local outlier factor of each sample.
+        """
+        norm_prob_local_outlier_factors = self.norm_factor * np.sqrt(
```

Comment:
I would inline the previous method here, it is overly verbose to make one method boilerplate per line of code.